### PR TITLE
Recent searches

### PR DESCRIPTION
**Pull request by tinhnguyen20**
[feature] Recent Searches - PivotalTrackerID #152020968 - completed
- recent searches runs robustly
- fixed race conditions with session variables

[feature] Cities MVC rework - PivotalTrackerID #152709121 - completed
- recent searches feature implemented using model associations
- added has_many City reference to users

[feature] Rspec Testing in Cities - PivotalTrackerID #152709186 - completed
- testing for recent_searches uses rspec due to racing issues with capybara/cucumber

Removed 
- Client_ID and Client Model functionality moved into Users by Jeffrey

heroku link: https://shrouded-brook-73438.herokuapp.com/

### CODE REVIEW

**Comment by alexberaut** [comment]
It looks like the branch can't be rebased because of conflicts, so this doesn't work. Also there should be a heroku link we can test, otherwise there is no way to confirm the app actually works and that you've rebased over my refactoring correctly.

**Review by ghost — CHANGES_REQUESTED** [review]
Could you explain or do these items?

* Can we have a heroku link for testing? There's no other way to confirm the app works.
* The branch can't be rebased due to conflicts, that needs to be fixed.
* Cant the Readme.Rdoc be removed? I replaced it yesterday with Readme.md. Also the Untitled document from the home directory; I forgot to delete it in my last PR.
* Can you explain or remove the commented out code in breathe_controller/.rb?
* Can you clear your public/assets folder and re run 'RAILS_ENV=development assets:precompile' ? looks like some extra/redundant asset files are getting added but I could be wrong. This could be happening because the filenames change on precompilation.
* Can you confirm that you can use the setup instructions on our readme to deploy this branch of the app? This could catch database and dependency issues that ruined my last week when I wasn't careful enough.
* Can you explain why the clients table was removed? Was nothing using it? I ask because I had a huge number of database bugs when debugging that got really painful to debug.

The implementation, testing, and documentation looks good!

**Review by deduq2 — APPROVED** [review]
LGTM

**Comment by ghost** [comment]
It still can't merge

File: app/controllers/cities_controller.rb [hunk 1]
```diff
@@ -36,49 +33,50 @@ def a_in_b_as_c?(a, b, c) # a in b as c
...
-    # Making a cookie to store all data of cities being searched.
-    def city_data
-      if params[:geo]
-        latlng = params[:geo]
-        loc_key = City.get_loc_key(latlng["lat"], latlng["lng"], params[:name])
-        city = City.find_by(location_key: loc_key)
+    # Helper method to reduce code complexity
+    def update_recent_cities(user, rec_cities)
+      if rec_cities.length > 5
+        rec_cities = rec_cities[rec_cities.length - 5, rec_cities.length - 1]
```
Comment (JeffreySpitz) [inline app/controllers/cities_controller.rb]: This could be `rec_cities = rec_cities.pop(5)`

File: app/controllers/cities_controller.rb [hunk 2]
```diff
@@ -36,49 +33,50 @@ def a_in_b_as_c?(a, b, c) # a in b as c
...
       end
-      city.update_city_data
-
+      if user.provider.nil?
+        user.update_attributes(recent_cities: rec_cities)
+        user.save(:validate => false)
+      else
+        user.update_attributes!(recent_cities: rec_cities)
+      end
+      user.recent_cities.reverse
```
Comment (JeffreySpitz) [inline app/controllers/cities_controller.rb]: I think `update_attributes` saves the user record so i don't think `user.save(:validate => false)` will help avoid the error.
Lines 41-46 could be replaced with `user.recent_cities = rec_cities` and  `user.save(:validate => false)`

File: app/models/city.rb [hunk 3]
```diff
@@ -23,28 +18,45 @@ def has_valid_data
...
     city = City.find_by(lat: lat, lng: lng)
     if city and !city.location_key.nil?
       return city.location_key
     end
     url = "http://dataservice.accuweather.com/locations/v1/cities/geoposition/search"
     query = {apikey: City.get_accuweather_key(), q: "#{lat},#{lng}",language:"en-us" }
-    response = City.get_resonse(HTTParty.get(url, query: query), url, query)
-    location_key = response["Key"]
+    resp = HTTParty.get(url, query: query)
+    location_key = resp["Key"]
```
Comment (JeffreySpitz) [inline app/models/city.rb:46]: Not DRY: replace lines 43-46 with `self.obtain_loc_key`

File: spec/controllers/cities_controller_spec.rb [hunk 4]
```diff
@@ -6,60 +6,66 @@
         before :each do
             @city = City.new(name: "Berkeley", lat: "37.8716", lng: "-122.2727", location_key: "332044")
             @city.save!
+            @city2 = City.new(name: "Fort Lauderdale", lat: "26.1224", lng: "-80.1373", location_key: "328168")
+            @city2.save!
+            @city3 = City.new(name: "")
```
Comment (JeffreySpitz) [inline spec/controllers/cities_controller_spec.rb]: Do you ever use @city3?

**Review by JeffreySpitz — CHANGES_REQUESTED** [review]
Looks good but i have left some minor code refactoring comments under Files Changed. Would also like to see a heroku link. Try using `git push heroku testbranch:master` https://devcenter.heroku.com/articles/git#deploying-code

**Review by MaadhavShah — COMMENTED** [review]
Main thing for me is also just the heroku link, can you push these changes to your own personal heroku and give us the link so that we can test that functionality works, especially for feature tests that we were unable to implement and have to manually test?

File: app/controllers/cities_controller.rb (hunk 2 above)
Comment (tinhnguyen20) [reply app/controllers/cities_controller.rb]: Noted! Will implement in next patch

File: app/controllers/cities_controller.rb (hunk 1 above)
Comment (tinhnguyen20) [reply app/controllers/cities_controller.rb]: Good tip. Will integrate

File: app/models/city.rb (hunk 3 above)
Comment (tinhnguyen20) [reply app/models/city.rb:46]: noted

File: spec/controllers/cities_controller_spec.rb (hunk 4 above)
Comment (tinhnguyen20) [reply spec/controllers/cities_controller_spec.rb]: Noted and deleted

**Comment by tinhnguyen20** [comment]
Made all the changes everyone suggested. initial comment includes heroku link that demonstrates the recent_searches functionality.

https://shrouded-brook-73438.herokuapp.com/

**Review by JeffreySpitz — APPROVED** [review]

**Comment by ghost** [comment]
Once this can be merged without conflicts then it looks ok.

**Pull request merged**